File: Leetcode/032/solution.java
// https://leetcode.com/problems/longest-valid-parentheses
// See the great solutions here: https://leetcode.com/articles/longest-valid-parentheses/
// Note that DP might not be the best solution for this problem
// For parentheses matching questions, think of the stack.
// Here's my solution: O(n) time, O(n) space.
class Solution {
    public int longestValidParentheses(String s) {
        int count = 0, maxLength = 0;
        int[] dp = new int[s.length() + 1];
        for (int i = 1; i < dp.length; i++) {
            if (s.charAt(i - 1) == '(') {
                count++;
            } else if (count > 0) {
                count--;
                dp[i] = 2 + dp[i-1];
                dp[i] = dp[i] + dp[i - dp[i]];
                maxLength = Math.max(dp[i], maxLength);
            }
        }
        return maxLength;
    }
}

File: Leetcode/032/solution.py
# https://leetcode.com/problems/longest-valid-parentheses
# See the great solutions here: https://leetcode.com/articles/longest-valid-parentheses/
# Note that DP might not be the best solution for this problem
# For parentheses matching questions, think of the stack.
# Here's my solution: O(n) time, O(n) space.
class Solution(object):
    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        count = max_len = 0
        dp = [0 for _ in range(len(s)+1)]
        for i in range(1, len(s) + 1):
            if s[i-1] == '(':
                count += 1
            elif count > 0:
                count -= 1
                dp[i] = 2 + dp[i-1];
                dp[i] = dp[i] + dp[i - dp[i]]
                max_len = max(dp[i], max_len)

        return max_len
